File: backend/models/maintenance.py
"""
Maintenance-related Pydantic models.

Contains models for maintenance requests, contractors, purchase orders, and invoices.
"""

from pydantic import BaseModel, ConfigDict, Field
from typing import List, Optional


class MaintenanceStatus:
    SUBMITTED = "submitted"
    UNDER_REVIEW = "under_review"
    APPROVED = "approved"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    REJECTED = "rejected"


class MaintenanceRequestCreate(BaseModel):
    title: str = Field(..., max_length=200)
    description: str = Field(..., max_length=2000)
    category: str = Field(..., max_length=100)  # plumbing, electrical, structural, common_area, lift, fire_safety, etc.
    location: str = Field(..., max_length=200)
    priority: str = Field("normal", max_length=20)  # low, normal, high, urgent
    images: List[str] = Field(default_factory=list, max_items=10)
    is_test_data: bool = False


class MaintenanceRequestResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    id: str
    title: str
    description: str
    category: str
    location: str
    priority: str
    status: str
    images: List[str] = Field(default_factory=list)
    submitted_by: str
    submitted_by_name: str
    assigned_contractor: Optional[str] = None
    contractor_name: Optional[str] = None
    estimated_cost: Optional[float] = None
    actual_cost: Optional[float] = None
    purchase_order_id: Optional[str] = None
    invoice_id: Optional[str] = None
    work_order_ids: List[str] = Field(default_factory=list)
    approval_history: List[dict] = Field(default_factory=list)
    notes: List[dict] = Field(default_factory=list)
    created_at: str
    updated_at: str
    completed_at: Optional[str] = None


class ContractorCreate(BaseModel):
    name: str = Field(..., max_length=200)
    abn: Optional[str] = Field(None, max_length=20)
    email: Optional[str] = Field(None, max_length=100)
    phone: Optional[str] = Field(None, max_length=20)
    address: Optional[str] = Field(None, max_length=300)
    services: List[str] = Field(default_factory=list, max_items=20)  # plumbing, electrical, cleaning, etc.
    notes: Optional[str] = Field(None, max_length=1000)
    insurance_expiry: Optional[str] = Field(None, max_length=100)
    public_liability_expiry: Optional[str] = Field(None, max_length=100)
    workers_comp_expiry: Optional[str] = Field(None, max_length=100)


class ContractorResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    id: str
    name: str
    abn: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    address: Optional[str] = None
    services: List[str]
    notes: Optional[str] = None
    insurance_expiry: Optional[str] = None
    public_liability_expiry: Optional[str] = None
    workers_comp_expiry: Optional[str] = None
    rating: Optional[float] = None
    total_jobs: int
    total_spent: float
    created_at: str


class PurchaseOrderCreate(BaseModel):
    maintenance_request_id: str = Field(..., max_length=50)
    contractor_id: str = Field(..., max_length=50)
    description: str = Field(..., max_length=1000)
    amount: float
    due_date: Optional[str] = Field(None, max_length=100)


class PurchaseOrderResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    id: str
    po_number: str
    maintenance_request_id: str
    contractor_id: str
    contractor_name: str
    description: str
    amount: float
    status: str  # draft, sent, accepted, completed, cancelled
    due_date: Optional[str] = None
    created_by: str
    created_at: str


class InvoiceCreate(BaseModel):
    purchase_order_id: str = Field(..., max_length=50)
    invoice_number: Optional[str] = Field(None, max_length=100)
    amount: float
    gst_amount: float = 0
    notes: Optional[str] = Field(None, max_length=1000)


class InvoiceResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    id: str
    invoice_number: Optional[str] = None
    purchase_order_id: Optional[str] = None
    po_number: Optional[str] = None
    contractor_id: Optional[str] = None
    contractor_name: Optional[str] = None
    amount: float
    gst_amount: float = 0.0
    total_amount: float = 0.0
    status: str  # pending, approved, paid
    notes: Optional[str] = None
    approved_by: Optional[str] = None
    approved_at: Optional[str] = None
    paid_at: Optional[str] = None
    created_at: str


__all__ = [
    "MaintenanceStatus",
    "MaintenanceRequestCreate",
    "MaintenanceRequestResponse",
    "ContractorCreate",
    "ContractorResponse",
    "PurchaseOrderCreate",
    "PurchaseOrderResponse",
    "InvoiceCreate",
    "InvoiceResponse",
]

File: frontend/src/types/maintenance.ts
export enum MaintenanceStatus {
    SUBMITTED = "submitted",
    UNDER_REVIEW = "under_review",
    APPROVED = "approved",
    IN_PROGRESS = "in_progress",
    COMPLETED = "completed",
    REJECTED = "rejected",
}

export interface MaintenanceApprovalEntry {
    action: "approved" | "rejected" | "submitted" | "assigned" | "completed" | "under_review";
    performed_by: string;
    performed_by_name?: string;
    timestamp: string;
    comment?: string;
}

export interface MaintenanceNote {
    id: string;
    content: string;
    created_by: string;
    created_by_name?: string;
    created_at: string;
}

export interface MaintenanceRequestResponse {
    id: string;
    title: string;
    description: string;
    category: string;
    location: string;
    priority: "low" | "normal" | "high" | "urgent";
    status: MaintenanceStatus;
    images: string[];
    submitted_by: string;
    submitted_by_name: string;
    assigned_contractor?: string | null;
    contractor_name?: string | null;
    estimated_cost?: number | null;
    actual_cost?: number | null;
    purchase_order_id?: string | null;
    invoice_id?: string | null;
    work_order_ids: string[];
    approval_history: MaintenanceApprovalEntry[];
    notes: MaintenanceNote[];
    created_at: string;
    updated_at: string;
    completed_at?: string | null;
}

export interface ContractorResponse {
    id: string;
    name: string;
    abn?: string | null;
    email?: string | null;
    phone?: string | null;
    address?: string | null;
    services: string[];
    notes?: string | null;
    insurance_expiry?: string | null;
    public_liability_expiry?: string | null;
    workers_comp_expiry?: string | null;
    rating?: number | null;
    total_jobs: number;
    total_spent: number;
    created_at: string;
}

export interface PurchaseOrderResponse {
    id: string;
    po_number: string;
    maintenance_request_id: string;
    contractor_id: string;
    contractor_name: string;
    description: string;
    amount: number;
    status: "draft" | "sent" | "accepted" | "completed" | "cancelled";
    due_date?: string | null;
    created_by: string;
    created_at: string;
}

export interface InvoiceResponse {
    id: string;
    invoice_number: string;
    purchase_order_id: string;
    po_number: string;
    contractor_id: string;
    contractor_name: string;
    amount: number;
    gst_amount: number;
    total_amount: number;
    status: "pending" | "approved" | "paid";
    notes?: string | null;
    approved_by?: string | null;
    approved_at?: string | null;
    paid_at?: string | null;
    created_at: string;
}
